packages/remote-replay-launcher/src/index.ts

```diff
@@ -49,6 +52,27 @@ export const executeRemoteTestRun = async ({
 
   const client = createClient({ apiToken });
 
+  if (appUrl && appDirectory) {
+    logger.error("You must provide precisely one of appUrl or appDirectory");
+    process.exit(1);
+  }
+
+  if (appDirectory) {
+    // If the app URL is not a valid URL, it must be a local directory and we bundle it and upload to S3
+    const { testRun } = await uploadAssetsToS3({
+      folder: appDirectory,
+      client,
+      commitSha,
+      rewrites: rewrites ?? [],
+    });
+    return { testRun };
```

Comment:
Note that by exiting early here none of the `onXX` callbacks will be called and the completion of the test run won't be awaited. Given how different the behaviour of the tunnel and non-tunnel variants is, it might be worth refactoring this to extract the common functionality into a separate function and the distinct logic into 2 functions ... or expose 2 functions and make the clients of this package call the correct one.

Either way it will be good to document that some of the options passed here are  incompatible (appDirectory & appUrl) (onTunnelCreated / isLockable / keepTunnelOpenPromise / onTunnelStillLocked / secureTunnelHost and rewrites)

Reply:
Yeah, I guess this made more sense before I addressed Daniel's comment of making appUrl/appDirectory separate parameters because then the caller could be agnostic to what it was passing through. But now it needs to handle the two parameters anyway, it's probably better to make it call two separate functions to make it clear they'll have different behaviour. Let me do that.